src/transformers/models/idefics2/modeling_idefics2.py

```diff
@@ -1473,16 +1473,23 @@ def __init__(self, config: Idefics2Config):
         super().__init__(config)
         self.padding_idx = self.config.text_config.pad_token_id
         self.vocab_size = self.config.text_config.vocab_size
+        self._use_flash_attention_2 = config._attn_implementation == "flash_attention_2"
 
         self.vision_model = Idefics2VisionTransformer(config.vision_config)
         self.connector = Idefics2Connector(config)
-        self.text_model = AutoModel.from_config(config.text_config)
-
+        text_model_kwargs = {}
+        if self._use_flash_attention_2:
+            text_model_kwargs["use_flash_attention_2"] = True
+            torch_dtype = None
+            if config.text_config.torch_dtype is not None:
+                torch_dtype = config.text_config.torch_dtype
+            elif config.torch_dtype is not None:
+                torch_dtype = config.torch_dtype
+            text_model_kwargs["torch_dtype"] = torch_dtype
+        self.text_model = AutoModel.from_config(config.text_config, **text_model_kwargs)
```

Comment:
* Let's use the same pattern as [in llava-next](https://github.com/huggingface/transformers/blob/569743f510780fc42aeb2364501584638f400558/src/transformers/models/llava_next/modeling_llava_next.py#L314). This is more robust to future attention implementations

* It's a bit funny to have two possible options for `torch_dtype` here, and taking the one from the text_config as precedence. If a user specified `model = IdeficsForConditionalGeneration(checkpoint, torch_dtype=torch.float16)` I'd expect `torch.float16` to be used.


```suggestion
        torch_dtype = config.text_config.torch_dtype
        if config.torch_dtype is not None:
            torch_dtype = config.torch_dtype 
        self.text_model = AutoModel.from_config(
            config.text_config, 
            attn_implementation=config.._attn_implementation, 
            torch_dtype=torch_dtype
        )
```